<a href="https://colab.research.google.com/github/jocinenabwire/MORINGA-WEEK-4-1P/blob/main/Moringa_WEEK_4_IP_2021_Jocine_Nabwire.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# **MORINGA PREP DATA SCIENCE INDEPENDENT PROJECT: WEEK 4**

In this week's independent project, you will be working as a data scientist working for an electric car-sharing service company. You have been tasked to process stations data to understand electric car usage over time

**Research Question**

Identify the most popular hour of the day for picking up a shared electric car (Bluecar) in the city of Paris over the month of April 2018.

**Bonus Questions (Optional)**

What is the most popular hour for returning cars?

What station is the most popular?
 
>>Overall?
 
>>At the most popular picking hour?

What postal code is the most popular for picking up Blue cars? Does the most popular station belong to that postal code?
 
>>Overall?
 
>>At the most popular picking hour?

Do the results change if you consider Utilib and Utilib 1.4 instead of Blue cars? 

**Datasets Provided**

1.Autolib_DDI_DB_description_MoringaSchool_w4 https://drive.google.com/file/d/13DXF2CFWQLeYxxHFekng8HJnH_jtbfpN/view

2.Autolib Dataset http://bit.ly/autolib_dataset

N/B:The dataset contains data collected for a period of 9 days.

**Hint:**

To compute usage, we will need to understand that we have to join successive (in time) measures/counters for a given station, as the difference will tell whether a car was picked up, returned, or nothing happened.

# **Importing Libraries**

In [1]:
# Importing the libraries we will need 

# Importing the pandas library
# 
import pandas as pd

# Importing the numpy library
#
import numpy as np


# **Reading the Dataset from our CSV file**

In [2]:
#loading the Autolib Dataset
df = pd.read_csv("http://bit.ly/autolib_dataset")

In [3]:
#Previewing the Autolib dataset
df.head()

Address  Cars  Bluecar counter  ...  day  hour  minute
0                 2 Avenue de Suffren     0                0  ...    8    11      43
1           145 Rue Raymond Losserand     6                6  ...    6     7      24
2    2 Avenue John Fitzgerald Kennedy     3                3  ...    3    20      14
3              51 Rue Eugène Oudiné     3                3  ...    4     4      37
4  6 avenue de la Porte de Champerret     3                3  ...    8    17      23

[5 rows x 25 columns]

In [4]:
#checking the tail
df.tail()

Address  Cars  Bluecar counter  ...  day  hour  minute
4995          8 avenue Ménelotte     2                2  ...    6    11      26
4996            37 rue de Dantzig     4                4  ...    4    16      56
4997               142 rue du Bac     1                1  ...    1     7       1
4998  2 avenue du Val de Fontenay     2                2  ...    4    17      27
4999          5 Rue Henry Brisson     4                4  ...    8    10       2

[5 rows x 25 columns]

# **Data Understanding**

In [5]:
#getting the info
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5000 entries, 0 to 4999
Data columns (total 25 columns):
 #   Column               Non-Null Count  Dtype 
---  ------               --------------  ----- 
 0   Address              5000 non-null   object
 1   Cars                 5000 non-null   int64 
 2   Bluecar counter      5000 non-null   int64 
 3   Utilib counter       5000 non-null   int64 
 4   Utilib 1.4 counter   5000 non-null   int64 
 5   Charge Slots         5000 non-null   int64 
 6   Charging Status      5000 non-null   object
 7   City                 5000 non-null   object
 8   Displayed comment    111 non-null    object
 9   ID                   5000 non-null   object
 10  Kind                 5000 non-null   object
 11  Geo point            5000 non-null   object
 12  Postal code          5000 non-null   int64 
 13  Public name          5000 non-null   object
 14  Rental status        5000 non-null   object
 15  Scheduled at         47 non-null     object
 16  Slots 

The dataset has 5000 entries and 25 columns.

In [6]:
#checking the columns
df.columns

Index(['Address', 'Cars', 'Bluecar counter', 'Utilib counter',
       'Utilib 1.4 counter', 'Charge Slots', 'Charging Status', 'City',
       'Displayed comment', 'ID', 'Kind', 'Geo point', 'Postal code',
       'Public name', 'Rental status', 'Scheduled at', 'Slots', 'Station type',
       'Status', 'Subscription status', 'year', 'month', 'day', 'hour',
       'minute'],
      dtype='object')

These are columns found in the dataset.

In [7]:
#Getting the shape
df.shape

(5000, 25)

The dataset has 5000 rows and 25 columns.

In [8]:
#getting basic statistical details
df.describe()

Cars  Bluecar counter  ...         hour       minute
count  5000.000000      5000.000000  ...  5000.000000  5000.000000
mean      2.333600         2.333600  ...    11.509200    29.270000
std       2.035274         2.035274  ...     6.893549    17.231741
min       0.000000         0.000000  ...     0.000000     0.000000
25%       0.000000         0.000000  ...     6.000000    14.000000
50%       2.000000         2.000000  ...    11.000000    29.000000
75%       4.000000         4.000000  ...    18.000000    44.000000
max       7.000000         7.000000  ...    23.000000    59.000000

[8 rows x 12 columns]

The count,mean,std,min,max and quantiles are shown above.

In [9]:
#getting unique values
df.nunique()

Address                1104
Cars                      8
Bluecar counter           8
Utilib counter            3
Utilib 1.4 counter        4
Charge Slots              3
Charging Status           3
City                     88
Displayed comment        25
ID                     1105
Kind                      4
Geo point              1105
Postal code             107
Public name            1105
Rental status             4
Scheduled at              8
Slots                     8
Station type              3
Status                    3
Subscription status       3
year                      1
month                     1
day                       9
hour                     24
minute                   60
dtype: int64

These shows the number of unique values in each column.

# **Data Cleaning**

## Validity

In [10]:
df.columns

Index(['Address', 'Cars', 'Bluecar counter', 'Utilib counter',
       'Utilib 1.4 counter', 'Charge Slots', 'Charging Status', 'City',
       'Displayed comment', 'ID', 'Kind', 'Geo point', 'Postal code',
       'Public name', 'Rental status', 'Scheduled at', 'Slots', 'Station type',
       'Status', 'Subscription status', 'year', 'month', 'day', 'hour',
       'minute'],
      dtype='object')

In [11]:
#Irrelevant Data
#Drop columns we do not need.
df.drop(['Address','Cars','Displayed comment', 'ID','Geo point'],axis=1,inplace=True)
df

Bluecar counter  Utilib counter  Utilib 1.4 counter  ...  day hour minute
0                   0               0                   0  ...    8   11     43
1                   6               0                   0  ...    6    7     24
2                   3               0                   2  ...    3   20     14
3                   3               1                   0  ...    4    4     37
4                   3               0                   0  ...    8   17     23
...               ...             ...                 ...  ...  ...  ...    ...
4995                2               0                   0  ...    6   11     26
4996                4               0                   0  ...    4   16     56
4997                1               0                   0  ...    1    7      1
4998                2               0                   0  ...    4   17     27
4999                4               0                   0  ...    8   10      2

[5000 rows x 20 columns]

## Completeness

In [12]:
#checking for null values across each column
df.isnull().sum()

Bluecar counter           0
Utilib counter            0
Utilib 1.4 counter        0
Charge Slots              0
Charging Status           0
City                      0
Kind                      0
Postal code               0
Public name               0
Rental status             0
Scheduled at           4953
Slots                     0
Station type              0
Status                    0
Subscription status       0
year                      0
month                     0
day                       0
hour                      0
minute                    0
dtype: int64

Scheduled at column has 4953 null values.

## Consistency

In [13]:
#checking for duplicates
df.duplicated()

0       False
1       False
2       False
3       False
4       False
        ...  
4995    False
4996    False
4997    False
4998    False
4999    False
Length: 5000, dtype: bool

There are no duplicates.

In [14]:
#rearranging our dataset so that the day and hour are arranged systematically.
df1=df.sort_values(['day','hour'], ascending=[True, True])
df1

Bluecar counter  Utilib counter  Utilib 1.4 counter  ...  day hour minute
123                 0               0                   0  ...    1    0     32
459                 3               0                   0  ...    1    0     37
584                 7               0                   0  ...    1    0      5
978                 2               0                   0  ...    1    0     33
1026                4               0                   0  ...    1    0     34
...               ...             ...                 ...  ...  ...  ...    ...
3899                3               0                   0  ...    9   23     24
4443                0               0                   0  ...    9   23     58
4582                5               0                   0  ...    9   23     56
4956                0               0                   0  ...    9   23     32
4992                0               0                   0  ...    9   23      1

[5000 rows x 20 columns]

From the Google Docs:

Total number of Blue Cars=7

Total number of Utilib=4

Total number of Utilib 1.4=5

Total number of charging slots=3

Total number of parking slots=7

In [15]:
#creating a new column which will contain the cumulative difference 
#for the Blue Car Counter
df1['Bluecar_Diff'] =df1['Bluecar counter'].diff()
df1.head(5)

Bluecar counter  Utilib counter  ...  minute  Bluecar_Diff
123                 0               0  ...      32           NaN
459                 3               0  ...      37           3.0
584                 7               0  ...       5           4.0
978                 2               0  ...      33          -5.0
1026                4               0  ...      34           2.0

[5 rows x 21 columns]

In [16]:
df1.tail()

Bluecar counter  Utilib counter  ...  minute  Bluecar_Diff
3899                3               0  ...      24           2.0
4443                0               0  ...      58          -3.0
4582                5               0  ...      56           5.0
4956                0               0  ...      32          -5.0
4992                0               0  ...       1           0.0

[5 rows x 21 columns]

Negative values in difference column stand for picking.

Postive values in difference colum stand for returning.

# **Answering Questions**

**Blue Car**

In [17]:
#Question 1
#Identify the most popular hour of the day for picking up a shared electric car 
#(Bluecar) in the city of Paris over the month of April 2018.
df1[df1['City'] =='Paris'][df1['Bluecar_Diff'] < 0].groupby('hour')['hour'].count().sort_values(ascending= False).head(3)


/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:4: UserWarning: Boolean Series key will be reindexed to match DataFrame index.
  after removing the cwd from sys.path.


hour
21    69
6     66
5     65
Name: hour, dtype: int64

21 is the most popular hour for picking up a shared electric car (Bluecar) in the city of Paris over the month of April 2018.




In [18]:
#Question 2
#What is the most popular hour for returning cars?
df1[df1['City'] =='Paris'][df1['Bluecar_Diff'] > 0].groupby('hour')['hour'].count().sort_values(ascending= False).head(3)

/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:3: UserWarning: Boolean Series key will be reindexed to match DataFrame index.
  This is separate from the ipykernel package so we can avoid doing imports until


hour
21    51
8     49
12    49
Name: hour, dtype: int64

21 is the most popular hour for picking up a shared electric car (Bluecar) in the city of Paris over the month of April 2018.

In [19]:
df1.head()

Bluecar counter  Utilib counter  ...  minute  Bluecar_Diff
123                 0               0  ...      32           NaN
459                 3               0  ...      37           3.0
584                 7               0  ...       5           4.0
978                 2               0  ...      33          -5.0
1026                4               0  ...      34           2.0

[5 rows x 21 columns]

In [20]:
#Question 3
#What station is the most popular?
df1[df1['City'] =='Paris'][df1['Kind']=='STATION'][df1['Status']=='ok'].groupby('Public name')['Public name'].count().sort_values(ascending=False).head(3)


/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:3: UserWarning: Boolean Series key will be reindexed to match DataFrame index.
  This is separate from the ipykernel package so we can avoid doing imports until


Public name
Paris/Porte de Montrouge/8    13
Paris/Voltaire/182            11
Paris/Courcelles/69           11
Name: Public name, dtype: int64

Paris/Porte de Montrouge/8 is the most popular station in Paris.

In [21]:
#Question 4
#Most popular station overally.
df1[df1['Kind']=='STATION'][df1['Status']=='ok'].groupby('Public name')['Public name'].count().sort_values(ascending=False).head(3)

/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:2: UserWarning: Boolean Series key will be reindexed to match DataFrame index.
  


Public name
Paris/Porte de Montrouge/8    13
Paris/Courcelles/69           11
Paris/Voltaire/182            11
Name: Public name, dtype: int64

Paris/Porte de Montrouge/8 is the most popular station overally.

In [25]:
#Question 5
#What postal code is the most popular for picking up Blue cars? 
df1[df1['City'] =='Paris'][df1['Bluecar_Diff']< 0].groupby('Postal code')['Postal code'].count().sort_values(ascending= False).head(3)

/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:2: UserWarning: Boolean Series key will be reindexed to match DataFrame index.
  
/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:2: UserWarning: Boolean Series key will be reindexed to match DataFrame index.
  


Postal code
75015    95
75008    92
75017    88
Name: Postal code, dtype: int64

75015 Postal Code is the most popular for picking up Blue Cars in Paris.  

**Utilib**

In [26]:
#creating a new column which will contain the cumulative difference 
#for the Utilib Counter
df1['Utilib_Diff'] =df1['Utilib counter'].diff()
df1.head(5)

Bluecar counter  Utilib counter  ...  Bluecar_Diff  Utilib_Diff
123                 0               0  ...           NaN          NaN
459                 3               0  ...           3.0          0.0
584                 7               0  ...           4.0          0.0
978                 2               0  ...          -5.0          0.0
1026                4               0  ...           2.0          0.0

[5 rows x 22 columns]

In [27]:
#Question 6
#Identify the most popular hour of the day for picking up Utilib 
#in the city of Paris over the month of April 2018.
df1[df1['City'] =='Paris'][df1['Utilib_Diff'] < 0].groupby('hour')['hour'].count().sort_values(ascending= False).head(3)


/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:4: UserWarning: Boolean Series key will be reindexed to match DataFrame index.
  after removing the cwd from sys.path.


hour
11    9
13    9
21    8
Name: hour, dtype: int64

11 is the most popular hour for picking up a Utilib in the city of Paris over the month of April 2018.

In [29]:
#Question 7
#What is the most popular hour for returning cars?
df1[df1['City'] =='Paris'][df1['Utilib_Diff']< 0].groupby('hour')['hour'].count().sort_values(ascending= False).head(3)

/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:3: UserWarning: Boolean Series key will be reindexed to match DataFrame index.
  This is separate from the ipykernel package so we can avoid doing imports until


hour
11    9
13    9
21    8
Name: hour, dtype: int64

11 is the most popular hour for returning up a Utilib in the city of Paris over the month of April 2018.

In [31]:
#Question 8
#What postal code is the most popular for picking up Blue cars? 
df1[df1['City'] =='Paris'][df1['Utilib_Diff']< 0].groupby('Postal code')['Postal code'].count().sort_values(ascending= False).head(3)

/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:3: UserWarning: Boolean Series key will be reindexed to match DataFrame index.
  This is separate from the ipykernel package so we can avoid doing imports until


Postal code
75015    19
75019    13
75018    11
Name: Postal code, dtype: int64

75015 Postal Code is the most popular for picking up Utilib Counter in Paris.

**Utilib 1.4**

In [32]:
#creating a new column which will contain the cumulative difference 
#for the Utilib 1.4 Counter
df1['Utilib 1.4_Diff'] =df1['Utilib 1.4 counter'].diff()
df1.head(5)

Bluecar counter  Utilib counter  ...  Utilib_Diff  Utilib 1.4_Diff
123                 0               0  ...          NaN              NaN
459                 3               0  ...          0.0              0.0
584                 7               0  ...          0.0              0.0
978                 2               0  ...          0.0              0.0
1026                4               0  ...          0.0              0.0

[5 rows x 23 columns]

In [33]:
#Question 9
#Identify the most popular hour of the day for picking up Utilib 
#in the city of Paris over the month of April 2018.
df1[df1['City'] =='Paris'][df1['Utilib 1.4_Diff'] < 0].groupby('hour')['hour'].count().sort_values(ascending= False).head(3)


/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:4: UserWarning: Boolean Series key will be reindexed to match DataFrame index.
  after removing the cwd from sys.path.


hour
6     17
3     17
16    14
Name: hour, dtype: int64

6 is the most popular hour for returning up a Utilib 1.4 in the city of Paris over the month of April 2018.

In [34]:
#Question 10
#What is the most popular hour for returning cars?
df1[df1['City'] =='Paris'][df1['Utilib 1.4_Diff']> 0].groupby('hour')['hour'].count().sort_values(ascending= False).head(3)

/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:3: UserWarning: Boolean Series key will be reindexed to match DataFrame index.
  This is separate from the ipykernel package so we can avoid doing imports until


hour
21    17
10    16
9     16
Name: hour, dtype: int64

21 is the most popular hour for picking a Utilib 1.4 in the city of Paris over the month of April 2018.

In [35]:
#Question 11
#What postal code is the most popular for picking up Blue cars? 
df1[df1['City'] =='Paris'][df1['Utilib 1.4_Diff']< 0].groupby('Postal code')['Postal code'].count().sort_values(ascending= False).head(3)

/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:3: UserWarning: Boolean Series key will be reindexed to match DataFrame index.
  This is separate from the ipykernel package so we can avoid doing imports until


Postal code
75015    28
75017    27
75016    21
Name: Postal code, dtype: int64

75015 Postal Code is the most popular for picking up Utilib Counter in Paris

# **Recommendations**



1.  Add more Blue Cars at the hour 21.
2.  Add more Utilib Cars at the hour 11.
3.  Add more Utilib 1.4 Cars at hour 6.
4.  Add more cars at the 75015 Postal Code since it is the most popular.
5.  The outcome changes when it comes to the Utilib and the Utilib 1.4.

